onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -2006,10 +2021,31 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<Node*>& fuse
         }
       }
 
-      // Run TRT inference
-      if (!trt_context->enqueueV2(&buffers[0], stream, nullptr)) {
-        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "TensorRT EP execution context enqueue failed.");
-      }
+      // Run TRT inference			  
+      if (trt_state->cuda_graph_enable)
+      {
+          if (*cuda_graph_ptr == nullptr) {
+            std::unique_ptr<cudaGraph_t> graph = std::make_unique<cudaGraph_t>();
+            *cuda_graph_ptr = trt_state->executable_cuda_graph->get();
+            //warm up to avoid capturing initialization
+            if (!trt_context->enqueueV2(&buffers[0], stream, nullptr)) {
+              return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, 
+                                     "TensorRT engine has operations that are not allowed in CUDA graph capture mode. ",
+                                     "Please disable trt_cuda_graph_enable.");
+            }
+            CUDA_CALL_THROW(cudaStreamBeginCapture(stream, cudaStreamCaptureModeRelaxed));
```

Comment:
If certain operators can't be supported by the TensorRT and CUDA EPs, does the CPU EP come into play ? If so, the same comment as this- https://github.com/microsoft/onnxruntime/pull/9978#r795356670

Reply:
The cuda graph capturing in this PR is only for TensorRT subgraphs. Unsupported ops can still fall back to CUDA/CPU EPs.

Reply:
But if some ops fall back to CPU EP, can we allow graph capture ? If I understand the documentation of CUDA graphs, CPU operations even if captured during the graph capture phase are elided in the replay (i.e.) the replay won't produce the right results, will it ?  

https://pytorch.org/docs/master/notes/cuda.html#constraints
`CPU work is not captured. If the captured ops include CPU work, that work will be elided during replay.`

Reply:
Understood now, thanks.

One question though: Let us say the model looks like this:   TensorRT subgraph 1 -> CPU op -> TensorRT subgraph 2 and you are capturing the graphs for both the TRT subgraphs, will the necessary synchronization logic before the CPU op happen even with the cuda graph setup ?

Reply:
I think the case is the same with and without cuda graph, isn't it? CPU needs to wait until TRT subgraph1 produces its result.  That's interesting if there is a better cuda graph setup that can avoid the wait.